utils/cache_ctx.go

```diff
@@ -0,0 +1,84 @@
+package utils
+
+// Borrowed from Osmosis
+// https://github.com/osmosis-labs/osmosis/blob/62757d309957fa9e02e6fb0b5dc8caf1ca68e696/osmoutils/cache_ctx.go
+// Wraps the execution of a function with a temporary context so that state changes can be discarded if an error occurs
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+	"runtime/debug"
+
+	"github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// This function lets you run the function f, but if theres an error or panic
+// drop the state machine change and log the error.
+// If there is no error, proceeds as normal (but with some slowdown due to SDK store weirdness)
+// Try to avoid usage of iterators in f.
+//
+// If its an out of gas panic, this function will also panic like in normal tx execution flow.
+// This is still safe for beginblock / endblock code though, as they do not have out of gas panics.
+func ApplyFuncIfNoError(ctx sdk.Context, f func(ctx sdk.Context) error) (err error) {
+	// Add a panic safeguard
+	defer func() {
+		if recoveryError := recover(); recoveryError != nil {
```

Comment:
If we use this same helper in the endBlocker I think you suggested that gas won't be a concern. Will that require a refactor of this block? If so, should we think about that now and implement it now that we're introducing this tool

Reply:
No need to refactor, this is setup to handle each error type:
* Normal errors go down [this](https://github.com/Stride-Labs/stride/pull/783/files/04f50a8bbf3274005a6cdbec9c862a886a6e64c3#diff-7e7821ad8a56b1dabae43f6a23e9eb0265cf8a5b763fc4793ed23ac3518c4fcdR39) branch
* Non-out-of-gas panic's go down [this](https://github.com/Stride-Labs/stride/pull/783/files/04f50a8bbf3274005a6cdbec9c862a886a6e64c3#diff-7e7821ad8a56b1dabae43f6a23e9eb0265cf8a5b763fc4793ed23ac3518c4fcdR32-R33) branch (panic is swallowed and converted to an error)
* Out-of-gas panic's go down [this](https://github.com/Stride-Labs/stride/pull/783/files/04f50a8bbf3274005a6cdbec9c862a886a6e64c3#diff-7e7821ad8a56b1dabae43f6a23e9eb0265cf8a5b763fc4793ed23ac3518c4fcdR28-R30) branch (actual panic is thrown)